Fix #1481: Separate EC2 template version fetching from template fetching

ishaanverma opened the pull request:
### Summary
> Describe your changes.

This change separates the retrieval of launch templates and their versions into distinct functions, resolving an issue where the `aws_handle_regions` decorator returns an empty list instead of a tuple, leading to a `ValueError` when unpacking.

I see that #845 specifically reverted a change that made `aws_handle_regions` polymorphic in return type.


### Related issues or links
> Include links to relevant issues or other pages.

- https://github.com/lyft/cartography/issues/...


### Checklist

Provide proof that this works (this makes reviews move faster). Please perform one or more of the following:
- [ ] Update/add unit or integration tests.
- [ ] Include a screenshot showing what the graph looked like before and after your changes.
- [ ] Include console log trace showing what happened before and after your changes.

If you are changing a node or relationship:
- [ ] Update the [schema](https://github.com/lyft/cartography/tree/master/docs/root/modules) and [readme](https://github.com/lyft/cartography/blob/master/docs/schema/README.md).

If you are implementing a new intel module:
- [ ] Use the NodeSchema [data model](https://cartography-cncf.github.io/cartography/dev/writing-intel-modules.html#defining-a-node).

achantavy commented on cartography/intel/aws/ec2/launch_templates.py:
```diff
@@ -21,27 +21,36 @@
+    template_versions: list[dict[str, Any]] = []
+    for template in templates:
+        template_id = template['LaunchTemplateId']
+        try:
```
Let's remove the try+catch here because @aws_handle_regions does this already

achantavy reviewed (changes requested):
Thank you so much for getting on this! I started it but didn't have time to complete the fix. This is exactly what I had in mind, we have one function to get templates and another to get template verisons. I left a comment about not needing to wrap with try + catch, let me know if you have questions.

I had it something like this:

```python
@timeit
@aws_handle_regions
def get_launch_templates(
    boto3_session: boto3.session.Session,
    region: str,
) -> list[dict[str, Any]]:
    client = boto3_session.client('ec2', region_name=region, config=get_botocore_config())
    paginator = client.get_paginator('describe_launch_templates')
    templates: list[dict[str, Any]] = []
    for page in paginator.paginate():
        templates.extend(page['LaunchTemplates'])
    return templates


@timeit
@aws_handle_regions
def get_launch_template_versions(
    boto3_session: boto3.session.Session,
    region: str,
) -> list[dict[str, Any]]:
    client = boto3_session.client('ec2', region_name=region, config=get_botocore_config())
    paginator = client.get_paginator('describe_launch_template_versions')
    versions: list[dict[str, Any]] = []
    for page in paginator.paginate():
        versions.extend(page['LaunchTemplateVersions'])
    return versions
```

ishaanverma commented:
ah, you're right. i can change it remove the try catch. although it's probably cleaner (perhaps faster) with your approach wherein `get_launch_template_versions` doesn't need the `templates` parameter and also doesn't rely on the function `get_launch_template_versions_by_template`. 

but do you think its possible that the AWS API call to fetch all `LaunchTemplateVersions` returns template ids that were not present when we fetch just the launch templates? would that cause issues when writing to the graph?

achantavy commented:
> but do you think its possible that the AWS API call to fetch all LaunchTemplateVersions returns template ids that were not present when we fetch just the launch templates? would that cause issues when writing to the graph?

Yeah, it is possible for getLTVs to return template IDs that were not previously present. If this is the case, the graph will look like

```
(:LaunchTemplateVersion)-->(:AWSAccount)
```

but it won't have this path

```
(:LaunchTemplateVersion)-->(:LaunchTemplate)
```

see https://github.com/cartography-cncf/cartography/blob/715e28b5fe964b53f73f7cecae1527534bdd4533/cartography/models/aws/ec2/launch_template_versions.py#L46 and https://github.com/cartography-cncf/cartography/blob/715e28b5fe964b53f73f7cecae1527534bdd4533/cartography/models/aws/ec2/launch_template_versions.py#L62.

This is ok.

ishaanverma commented:
gotcha, that makes sense. i've refactored the function, also removed the `get_launch_template_versions_by_template` function since it doesn't seem to be used anymore

achantavy reviewed (approved):
Looks good to me, thank you for your help!
